packages/salesforce-adapter/src/custom_object_instances_deploy.ts

```diff
@@ -0,0 +1,179 @@
+/*
+*                      Copyright 2020 Salto Labs Ltd.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import _ from 'lodash'
+import { values } from '@salto-io/lowerdash'
+import {
+  ChangeGroup, getChangeElement, DeployResult, Change,
+  InstanceElement, isAdditionGroup, isRemovalGroup, isModificationGroup, isInstanceChange,
+  ChangeGroupId, ModificationChange, RemovalChange, AdditionChange,
+} from '@salto-io/adapter-api'
+import {
+  isInstanceOfCustomObject, instancesToCreateRecords, apiName, instancesToDeleteRecords,
+  instancesToUpdateRecords,
+} from './transformers/transformer'
+import SalesforceClient from './client/client'
+import * as constants from './constants'
+import { Filter } from './filter'
+
+const { isDefined } = values
+
+const validateSingleTypedGroup = (changeGroup: {
+  groupID: ChangeGroupId
+  changes: ReadonlyArray<Change<InstanceElement>>
+}): void => {
+  const instances = changeGroup.changes.map(change => getChangeElement(change))
+  if (instances.length === 0) {
+    return
+  }
+  const instancesType = apiName(instances[0].type)
+  if (!instances.every(instance => apiName(instance.type) === instancesType)) {
+    throw new Error('Custom Object Instances change group should have a single type')
+  }
+}
+
+const deployAdditionGroup = async (
+  changeGroup: ChangeGroup<AdditionChange<InstanceElement>>,
+  client: SalesforceClient,
+  filtersRunner: Required<Filter>,
+): Promise<DeployResult> => {
+  validateSingleTypedGroup(changeGroup)
+  const instances = changeGroup.changes.map(change => getChangeElement(change))
```

Comment:
edit: there is a comment later in this file that might make this comment irrelevant, read the other comments before doing anything with this one

a few nits:
1. `const instances = changeGroup.changes.map(change => getChangeElement(change))` is duplicated (runs in `validateSingleTypedGroup` and here). the type name is also needed in both functions (calling `apiName` on instance types)
2. in `validateSingleTypedGroup` there is some treatment for the case where `instance.length === 0`, but here it will just crash in the next line. not sure if we need to handle that case at all, but the inconsistency makes it strange IMO
3. maybe we can address both of these issues at once with a small refactor:

```suggestion
  const instances = changeGroup.changes.map(change => getChangeElement(change))
  const instanceTypes = [...new Set(instances.map(inst => apiName(inst.type)))]
  if (instanceTypes.size !== 1) throw new Error(`... only one type allowed but got: ${instanceTypes}`)
```